MNT New helper for effective number of OpenMP threads

jeremiedbb opened the pull request:
some PRs are waiting for a helper to go from `n_jobs` to the number of threads to set in `prange` (which does not accept None or negative values). (see #13264)

This PR proposes an equivalent of joblib.effective_n_jobs but with a different behavior when `n_jobs = None`. In that case OpenMP can use as much threads as possible. It's done using both `omp_get_max_threads` and `joblib.effective_n_jobs` because loky's `effective_n_jobs` is more clever than just returning the cpu_count.

Not sure about the name though.

cc @ogrisel @tomMoral

jeremiedbb committed:
openmp helper equivalent of effective_n_jobs

rth commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,26 @@
+    if n_threads == 0:
+        raise ValueError("n_threads = 0 is invalid")
+
+    max_threads = min(openmp.omp_get_max_threads(), effective_n_jobs(-1))
```
Is there a case when `openmp.omp_get_max_threads()` would be lower than `joblib.effective_n_jobs(-1)` ?

jeremiedbb replied:
yes, if a user has first set `OMP_NUM_THREADS=1` for instance

tomMoral replied:
Also, some implementation of OpenMP return the number of __physical cores__ instead of virtual cores, which is typically lower than what is reported by `joblib.effective_n_jobs`.

ogrisel replied:
Interesting. This does not always seem to be the case though. Especially not libgomp. So it would be great if loky could have an option to report the number of physical cores as commented here :https://github.com/scikit-learn/scikit-learn/pull/14196#pullrequestreview-290041629

jeremiedbb committed:
protect openmp calls

jeremiedbb commented:
@NicolasHug you might also be interested in this.

tomMoral commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,36 @@
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
```
Maybe mention the fact that `omp_get_nax_threads` lookup to `OMP_NUM_THREADS`.

tomMoral commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,36 @@
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
+    - For ``n_threads > 0``, use this as the maximal number of threads for
+      parallel OpenMP calls.
```
Are we sure that we allow for over-subscription? Can't we raise a warning if `n_threads` is larger than `max_threads` and return `max_threads`?

jeremiedbb replied:
@ogrisel suggested so.

ogrisel replied:
I think it's fine to allow for oversubscription without warning as long as the default setting does not cause oversubscription. For some algorithms with a lot of IO a bit of oversubscription can actually be beneficial.

tomMoral commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,36 @@
+        raise ValueError("n_threads = 0 is invalid")
+
+    IF SKLEARN_OPENMP_SUPPORTED:
+        max_threads = min(openmp.omp_get_max_threads(), effective_n_jobs(-1))
```
`max_n_threads`?

jeremiedbb committed:
comment openmp max threads

jeremiedbb committed:
right place comment

NicolasHug commented:
(Oops sorry, wrong button!)

I'm confused by the fact that the default (None) means "use as many threads as possible" when the parameter relates to OpenMP, but means "only use 1 thread (unless backend is specified...)" when it relates to joblib.
I think it's going to be hard to explain.

I'm also not sure whether it's right to use the same parameter name "n_jobs" for both OpenMP and joblib. Why not having `n_omp_threads`, and `n_jobs`? A "job" really is a joblib concept, not an OpenMP concept. OpenMP only deals with threads.

jeremiedbb commented:
> I'm confused by the fact that the default (None) means "use as many threads as possible" when the parameter relates to OpenMP, but means "only use 1 thread (unless backend is specified...)" when it relates to joblib.
I think it's going to be hard to explain.

The idea is that we let the BLAS use as many threads as possible, so why not do it also for our low level multi-threaded parts (by default).

> I'm also not sure whether it's right to use the same parameter name "n_jobs" for both OpenMP and joblib. Why not having n_omp_threads, and n_jobs? A "job" really is a joblib concept, not an OpenMP concept. OpenMP only deals with threads.

It would be less user friendly to have estimators with n_jobs and estimators with n_omp_threads. However It would be confusing that n_jobs=None does not have the same behavior for all estimators.
I'm unhappy in both cases :(

NicolasHug commented:
> It would be less user friendly to have estimators with n_jobs and estimators with n_omp_threads

I feel the opposite: it would give them more control, and they would know what `n_jobs` refers to. Hiding too much of the logic can also be detrimental to ease of use, sometimes.

If we support both `n_jobs` and `n_omp_threads`, the default for `n_omp_threads` could be 'auto', which would mean "use as many threads as possible, avoiding over-subscription". E.g. with 8 cores, a GridSearch with n_jobs=4 would result in `n_omp_threads` set to 2 for the underlying estimator. If a user sets `n_omp_threads` to an int, then they might risk over/under subscription (but that's their problem).

tomMoral commented:
The drawback of using multiple kw arguments is that it makes it less
consistent to use and to discover in the user API.
I would rather see some good documentation about the underlying parallelism
for each docstring with `n_jobs` to explain the default behavior.

Also, we discussed with @ogrisel about changing the default behavior for
`n_jobs=None` with the threading backend to the maximal number of threads,
which would make it easier to explain (thread based parallelism vs process
based parallelism behavior).

I agree changing `n_jobs` to something like `n_workers` would be
semantically better (even for joblib) but I guess it is very complicated.

Le mer. 26 juin 2019 à 19:07, Nicolas Hug <notifications@github.com> a
écrit :

> It would be less user friendly to have estimators with n_jobs and
> estimators with n_omp_threads
>
> I feel the opposite: it would give them more control, and they would know
> what n_jobs refers to. Hiding too much of the logic can also be
> detrimental to ease of use, sometimes.
>
> If we support both n_jobs and n_omp_threads, the default for n_omp_threads
> could be 'auto', which would mean "use as many threads as possible,
> avoiding over-subscription". E.g. with 8 cores, a GridSearch with n_jobs=4
> would result in n_omp_threads set to 2 for the underlying estimator. If a
> user sets n_omp_threads to an int, then they might risk over/under
> subscription (but that's their problem).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/14196?email_source=notifications&email_token=AAZKZ6N744RLMTYKD4ZP3ILP4OO4ZA5CNFSM4H3RVYB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYUF73Y#issuecomment-505962479>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAZKZ6OSETDBJBCB44XJ2STP4OO4ZANCNFSM4H3RVYBQ>
> .
>

jeremiedbb commented:
> Also, we discussed with @ogrisel about changing the default behavior for `n_jobs=None` with the threading backend to the maximal number of threads, which would make it easier to explain (thread based parallelism vs process based parallelism behavior).

If this change happens in joblib, it would definitely make sense to have the same behavior for OpenMP based parallelism.

NicolasHug commented:
But the default backend is loky right? So for `n_jobs` to have the same behaviour in joblib and openmp, users would have to wrap their call with the joblib context manager and use the threading backend?

rth commented:
> But the default backend is loky right? So for n_jobs to have the same behaviour in joblib and openmp, users would have to wrap their call with the joblib context manager and use the threading backend?

There are a few places where we currently use `prefer="threads"` (cf `neighbours.base`, `forest`, `iforest`, `logistic`, `linear_model.coordinate_descent` modules), so the use of `n_jobs` currently doesn't say anything about whether threading or loky backend is used.

rth commented:
Also there are places where we now use the loky multiprocessing backend where we might want to switch to the threading joblib backend or OpenMP in the future. It would be a bit akward if we have to deprecate `n_jobs` in favour of `n_threads` (or vice versa) to do that for an estimator (or have to revert if we change our mind about the parallel backend yet again later). 

The way parallelism is done for an estimator is an implementation detail that should be mostly transparent to the user.

NicolasHug commented:
> The way parallelism is done for an estimator is an implementation detail that should be mostly transparent to the user.

I don't disagree. But then the default should be consistent.

tomMoral commented:
> I don't disagree. But then the default should be consistent.

But it is hard to decide with which rule you should be consistent:

1 -  By default, C-level threadpools use all cores, so for thread base parallelism, we could use the same
2 - By default, we want `joblib` process based parallelism to be deactivated (as it might hurt the performance in some cases) so it should be the same for thread based parallelism.

The rational behind following rule 1 is that using all thread tends to always improve the performances, so we should use it (as it is done in `numpy`).

rth commented:
This is currently blocking us from adding OpenMP code and I think we should merge it, it's just a private function. There is still some time to decide how the public API should look like for this. 

Is it still a WIP @jeremiedbb ?

For OpenMP maybe it could make sense to even not expose `n_jobs` (or an equivalent parameter), and just have it enabled by default same as BLAS and numpy does. Because in practice, if `n_jobs=None` means `-1` (and we are handling oversubscription issues), maybe setting limiting it via an env variable at startup could be enough.

jeremiedbb commented:
The WIP tag was until there's a consensus :)

> it's just a private function. There is still some time to decide how the public API should look like for this.

I agree

> For OpenMP maybe it could make sense to even not expose n_jobs (or an equivalent parameter), and just have it enabled by default same as BLAS and numpy does. Because in practice, if n_jobs=None means -1 (and we are handling oversubscription issues), maybe setting limiting it via an env variable at startup could be enough.

I'm in favor of exposing `n_jobs` with a good documentation for it (Nicolas opened an issue about that).
I think there's value in keeping control of parallelism at runtime. For instance you might want to set n_jobs to -1 when fitting an estimator alone but set it to 1 when you put it into a gridsearch, without having to restart ipython.

However this can be decided later outside of this PR, maybe we should open a dedicated issue so that others can give their opinion.

rth commented:
> The WIP tag was until there's a consensus :)

For me WIP tags means that the author hasn't finished work or is planning to add more things. So if you are asking for feedback you should remove it IMO. 

> However this can be decided later outside of this PR, maybe we should open a dedicated issue so that others can give their opinion.

+1 for opening a issue about it listing possible options.

rth reviewed (approved):
Since it's only a private helper. cf the discussion on the public API in https://github.com/scikit-learn/scikit-learn/issues/14265#issuecomment-508745883

jeremiedbb committed:
don't import joblib if unecessary

NicolasHug commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,37 @@
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
```
I'm not sure I understand the purpose of also checking `joblib.effective_n_jobs(-1)`?

Is this for cases where you want to run code in a joblib context manager?

jeremiedbb replied:
It's because effective_n_jobs is more clever than cpu_count or omp_get_max_threads is some cases. I don't recall in which situations but @tomMoral should now better.

NicolasHug replied:
Ok, would be nice to have a comment explaining this then

tomMoral replied:
The `loky` implementation of `cpu_count`  better handles containers and CPU affinity. (see the docstring [here](https://github.com/tomMoral/loky/blob/master/loky/backend/context.py#L104) for detailed description).


However, it might be better to call `joblib.cpu_count` here instead of `effective_n_jobs`, as it also calls `loky.cpu_count` under the hood, .

NicolasHug replied:
Reading the docstring, setting `LOKY_MAX_CPU_COUNT` will affect the output of `_openmp_effective_n_threads()`.

I don't think that makes sense?

tomMoral replied:
I don't think this is a problem and it is necessary for the `loky` test suite.

NicolasHug replied:
but that's a loky variable that influences openmp behaviour?

tomMoral replied:
Its a global variable to limit the number of usable CPU in application that uses `loky`. There is a low change that it is used by anyone. Even if it is set, it would make sense to also limit the number of CPU used in `openmp` as this limit is considered as a global limit on the number of CPU.

NicolasHug replied:
>Its a global variable to limit the number of usable CPU in application that uses loky.

So it has nothing to do with openmp, and it's only remotely related to scikit-learn. I don't think we should rely on this.

tomMoral replied:
I agree that it would be better it is not ideal to have this environment variable impact `openmp` usage in `scikit-learn`.

However, `sklearn.utils.cpu_count` (which is `joblib.cpu_count`) seems to be the way advocated for in the doc to get the number of CPU. So I don't think we should change this in this PR. We can open a discussion about this in another PR.

NicolasHug replied:
It is only used in a single place though, in `test_multi_output_classification_partial_fit_parallelism`. None of the estimators / tools rely on it.

Also the sklearn version is deprecated.

tomMoral replied:
Indeed, it the `sklearn.utils.cpu_count` is deprecated in favor of `joblib.cpu_count`, sorry for the mistake.

Note that it is used under the hood by all calls to `Parallel`, so it is still the main way to compute the number of CPU in `scikit-learn`. I still don't think we should re-implement something in this PR, but we can discuss this in a PR in `joblib` to see if we can avoid this environment variable.

NicolasHug replied:
Ok, maybe Im overthinking it. I'm interested in @ogrisel's opinion on this

jeremiedbb replied:
What I wanted to do is just a conservative way to find the max number of cores to use. I think that `omp_get_num_procs` could overestimate it in some situations (e.g. containers although I'm not sure about that). A more robust method is joblib's effective_n_jobs. I think you should not see this as controlling nthreads through joblib but instead use joblib to gather more informations in order to set the most appropriate number of threads.

jeremiedbb replied:
On the other hand I agree that the fact that it depends onthe backend is a bit brittle. Tom's proposition is better.

ogrisel replied:
I like the fact that we do not use 32 openmp threads by default when running in a docker container that can only run with a 2000 mCPU quota in the cgroups. This is typically what we have on CI systems such as travis or circle: we run in docker containers with CPU quota but `openmp.omp_get_max_threads()` would naively return 32 leading to sever oversubscription that causes a significant slow down of the computation.

@NicolasHug so you would rather duplicate the code from loky just to remove the dependency of the `LOKY_MAX_CPU_COUNT` env variable? We could do that but that feels a bit overkill. I don't expect many people to mess around with `LOKY_MAX_CPU_COUNT`.

jnothman commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,37 @@
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
+      The result of ``omp_get_max_threads`` can be influenced by environement
```
```suggestion
      The result of ``omp_get_max_threads`` can be influenced by environment
```

jnothman reviewed (commented):
Implementation lgtm.

tomMoral commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,37 @@
+IF SKLEARN_OPENMP_SUPPORTED:
+    cimport openmp
+    from joblib import effective_n_jobs
```
```suggestion
    from joblib import cpu_count
```

tomMoral commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,37 @@
+        raise ValueError("n_threads = 0 is invalid")
+
+    IF SKLEARN_OPENMP_SUPPORTED:
+        max_n_threads = min(openmp.omp_get_max_threads(), effective_n_jobs(-1))
```
```suggestion
        max_n_threads = min(openmp.omp_get_max_threads(), cpu_count())
```

As discussed with @NicolasHug, this function should not be impacted by joblib backends and context so it is better to rely directly on `cpu_count`.

tomMoral commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,37 @@
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
```
```suggestion
      openmp.omp_get_max_threads() and joblib.cpu_count().
```

ogrisel commented:
Before merging this I would like to do some benchmarks on machines with 2 NUMA nodes or more. I suspect that in some cases it's actually detrimental to use all the cores but I would really like to know how bad it is on a bunch of use representative use cases (e.g, HistGradientBoostingClassifier and the new KMeans).

ogrisel commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,37 @@
+    - For ``n_threads > 0``, use this as the maximal number of threads for
+      parallel OpenMP calls.
+    - For ``n_threads < 0``, use the maximal number of threads minus
+      ``|n_threads + 1|``.
```
Maybe expand: "In particular `n_threads=-1` will use as many threads as there are available cores on the machine.".

jeremiedbb commented:
Here's a benchmark for KMeans (gbt coming soon).  
It's on a 88 cores machine (44 physical + hyperthreads) with 2 numa nodes.

![image](https://user-images.githubusercontent.com/34657725/65053737-b6517d80-d96c-11e9-89ff-04bf37ca2113.png)

For KMeans it seems there's no scalability issue w.r.t numa nodes.

However, using hyperthreads is bad. In this plot only the last orange point uses hyperthreads.

jeremiedbb commented:
Here's a benchmark of HistGradientBoostingClassifier. Same machine.

![image](https://user-images.githubusercontent.com/34657725/65057907-a5f0d100-d973-11e9-89c4-1edac190ae8d.png)

Scalability is not great, but it does not seem to be related to numa nodes.

Using hyperthreads is even worse.

ogrisel reviewed (approved):
LGTM.

We can improve `joblib/loky.cpu_count()` to add an option such as `loky.cpu_count(include_hyperthreads=False)` but this can come later.

NicolasHug commented:
Can we just add a comment saying that since we're using `joblib.cpu_counts` then `LOKY_MAX_CPU_COUNT` will also affect the output?

ogrisel commented:
Done. I have also updated the KMeans PR #11950 to use this branch and it seems to work as expected.

NicolasHug commented:
I guess we need to agree on https://github.com/scikit-learn/scikit-learn/issues/14265 now

rth commented:
Given the consensus in https://github.com/scikit-learn/scikit-learn/issues/14265 this is still useful, isn't it?

jeremiedbb commented:
yes

rth commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,41 @@
+        elif n_threads < 0:
+            return max(1, max_n_threads + n_threads + 1)
+
+        return n_threads
```
We could have refactored to avoid unnecessary syscalls when n_threads is give, but since we are always going to use it with `n_threads=None` it should not matter.

```py
if n_threads is not None and n_threads > 1:
    return n_threads

max_n_threads = min(openmp.omp_get_max_threads(), cpu_count())

if n_threads is None:
    return max_n_threads
elif n_threads < 0:
    return max(1, max_n_threads + n_threads + 1)
else:
    # should not happen
    raise ValueError
```

rth reviewed (approved):
This still look good. Given the previous discussion this is going to be used with `n_threads=None` by default mainly, but it's good to keep an explicit `n_threads` parameter for bechmarking during development.

@NicolasHug any objections to merge?

NicolasHug commented:
Sorry but now I fail to see where and how this helper will be used?

jeremiedbb commented:
Maybe in hgbt since it does not seem good to let all cores by default given the scalability issue.
KMeans #11950 will also need it if we don't deprecate the n_jobs parameter.

rth commented:
By default OpenMP will use all available cores. However for instance inside Docker in CircleCI the hardware has 32 cores but only 2 are actually available This maxes it out the thread count at `joblib.cpu_count()` additionally, which would address that by only using actually available cores with CPU affinity. In the future it could avoid using hyper-threaded cores as well. Basically this leave the flexibility to implement smarter detection of available cores in loky.

rth commented:
Though an issue with this PR is that it will ignore `OMP_NUM_THREADS` if `loky.cpu_count()` is lower. IMO `OMP_NUM_THREADS` should still have the highest priority.

NicolasHug commented:
If we start using this, doesn't this make the docs proposed in https://github.com/scikit-learn/scikit-learn/pull/15116 wrong now??

rth commented:
> If we start using this, doesn't this make the docs proposed in #15116 wrong now??

Not as far as I can tell. 

The sentence under `OpenMP-based parallelism`

> By default (and unless joblib is trying to
avoid oversubscription), the implementation will use as many threads as
possible.

is still correct, as long as we keep `OMP_NUM_THREADS` with the highest priority as compared to loky etc.

NicolasHug commented:
Ok so that still works because "as many thread as possible" is actually quite vague and doesn't really mean anything...


> However for instance inside Docker in CircleCI the hardware has 32 cores but only 2 are actually available

I'm not super familiar with all this, could you briefly explain (or link somewhere) how this may happen? Basically, who is at fault here? Shouldn't the docker container only expose 2 CPUs (Not even sure my question makes sense)?

jeremiedbb commented:
> Though an issue with this PR is that it will ignore OMP_NUM_THREADS if loky.cpu_count() is lower. IMO OMP_NUM_THREADS should still have the highest priority.

if nthreads is None we want to use at most as many threads as there are cores. So I think keeping joblib.cpu_count as an upper bound is a good thing.

Besides, prange won't use more threads than the number of cpus (guessed by openmp) and same for mkl, openblas,..

rth commented:
> I'm not super familiar with all this, could you briefly explain (or link somewhere) how this may happen? Basically, who is at fault here? Shouldn't the docker container only expose 2 CPUs (Not even sure my question makes sense)?

No one is at fault. Basically there are multiple mechanisms to limit CPU ressources for a process: [CPU affinity is one](https://en.wikipedia.org/wiki/Processor_affinity),  [linux CSF scheduler is another](https://en.wikipedia.org/wiki/Completely_Fair_Scheduler). Docker is one place where those are used, but not the only one (can also encounter those in other virtualized environments). OpenMP is oblivious to those, while loky/joblib should handle them and return the [correct number of available CPU cores](https://github.com/tomMoral/loky/blob/dc2d941d8285a96f3a5b666a4bd04875b0b25984/loky/backend/context.py#L105).

rth commented:
> if nthreads is None we want to use at most as many threads as there are cores. So I think keeping joblib.cpu_count as an upper bound is a good thing.
> Besides, prange won't use more threads than the number of cpus (guessed by openmp) and same for mkl, openblas,..

My concerns is more in the case loky limits the number of cores for some reason (say you are inside a parallel process), then no mater what you set as `OMP_NUM_THREAD` on CLI it will have no effect, which is somewhat counter-intuitive. Also it's much easier to explain that you can always force the number of threads manually with `OMP_NUM_THREADS` , as opposed to some complicated logic inside loky.

ogrisel commented:
> I'm not super familiar with all this, could you briefly explain (or link somewhere) how this may happen? Basically, who is at fault here? Shouldn't the docker container only expose 2 CPUs (Not even sure my question makes sense)?

docker typically sets CPU usage quotas via the [cgroups](https://en.wikipedia.org/wiki/Cgroups) feature of the Linux kernel. Typically OpenMP runtimes do not introspect cgroups quotas and will think they can use 32 concurrent threads while the Linux scheduler will ensure that only 2 can run at the same time, therefore causing oversubscription. `loky.cpu_count()` in smart enough to introspect such quotas (along with other constraints such as OS scheduler CPU affinity masks):

https://github.com/tomMoral/loky/blob/master/loky/backend/context.py#L104-L153

tomMoral commented:
> My concerns is more in the case loky limits the number of cores for some reason (say you are inside a parallel process), then no mater what you set as `OMP_NUM_THREAD` on CLI it will have no effect, which is somewhat counter-intuitive. Also it's much easier to explain that you can always force the number of threads manually with `OMP_NUM_THREADS` , as opposed to some complicated logic inside loky.

If you are in parallel processes, `OMP_NUM_THREADS` will have precedence over `joblib` limits. The limit is only changed when no other mechanism is changing it.

Note that `loky.cpu_count` is a global value that is never changed by loky. It only depends on the system ('# cores', affinity, ...)

ogrisel commented:
I believe that @rth's comment is about the value returned by `_openmp_effective_n_threads`:

if `OMP_NUM_THREADS` is set` it should take precedence over-what is computed by `loky.cpu_count()`.

tomMoral commented:
Sorry, I got confused with the parallel process. I think it does not change the behavior, no?

To me the question is simply, should we allow the user to go over `cpu_count` by setting `OMP_NUM_THREADS`? OpenBLAS and MKL seem to limit to `cpu_count` while OpenMP is not. I think it is better to do the former but I am fine with both solution.

jeremiedbb commented:
Current behavior is respect `OMP_NUM_THREADS` unless it exceeds `joblib.cpu_count`.
I think it's a good behavior and not so hard to document: OpenMP behavior, bounded by number of cpus.

I don't like making OMP_NUM_THREADS top priority because it's not in openmp specs.
The prority chain is:
- `OMP_NUM_THREADS` (lowest)
- `omp_set_num_threads` (at runtime)
- `prange(n, num_threads=num_threads) (as a prange arg)

So if the goal is to use this helper to set the number of threads as a prange argument, we should not make `OMP_NUM_THREADS` more of a priority than `omp_set_num_threads`.

NicolasHug commented on sklearn/utils/openmp_helpers.pyx:
```diff
@@ -0,0 +1,41 @@
+cpdef _openmp_effective_n_threads(n_threads=None):
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
```
We should always pass `n_threads=None` in the code right? The rest should only be for debugging?

jeremiedbb replied:
Most of the time yes.

Few exceptions : if we keep n_jobs in kmeans we would pass it to this function. If we decide to add a parameter to histgradientbosstingtrees to control the number of threads we would pass it to.

NicolasHug replied:
wait, we agreed in https://github.com/scikit-learn/scikit-learn/issues/14265 not to expose a parameter for openmp threads. We should deprecate `n_jobs` for Kmeans and link to the doc from #15116 now right?

jeremiedbb replied:
The openmp implement of kmeans is not merged yet so it's not yet needed.

We agreed to do that case by case. For most situations we agreed to expose nothing but we also agreed that we can discuss it for estimators where the main loop is parallelized with openmp.

thomasjpfan replied:
What do you think of `sklearn.set_config(n_omp_threads=...)`, which can be used directly in `prange(..., num_threads)`? The highest python level caller could get the value and pass it down into the cython function.

NicolasHug replied:
> We agreed to do that case by case

Really? Where? My understanding is that we went for option 5.

Going case by case (i.e. accepting n_jobs for controlling openMP in some cases) is going to be extremely painful to explain to users.  #15116  is already a lot to digest. We need to keep things simple

NicolasHug replied:
> What do you think of sklearn.set_config(n_omp_threads=...),

I think that we agreed on 5 and we should stick with 5. This is not the time for new ideas... That time was for #15116 

(but to address your solution, I don't see how this is different from what setting OMP_NUM_THREADS)

jeremiedbb replied:
> Really? Where? My understanding is that we went for option 5.

during the last meeting. We went for option 5. But we allowed to keep the discussion opened on a case by case. It does not mean we won't use 5 everywhere :)

NicolasHug commented:
Thanks @rth @ogrisel I see why we want the helper now.

I think I agree with @rth that `OMP_NUM_THREADS` must have top priority here:

- That gives users a simple way to specify "I know what I'm doing and I want *that* many threads no matter what anybody else says". Also we clearly say in #15116 that setting `OMP_NUM_THREADS` has the highest priority
- How can joblib protect from oversubscription if `OMP_NUM_THREADS` isn't respected?

jeremiedbb commented:
The implemented behavior is `omp_get_max_threads`, bounded by `joblib.cpu_count`.

the usage is
```
num_threads = _openmp_effective_n_threads(nthreads)

for i in prange(N, num_threads=num_threads)
```

When doing it this way OpenMP will use `num_threads` and no env var or function can change that: setting the number of threads in prange args has the highest priority.

But `num_threads` is actually controlled by those. Indeed, `omp_get_max_threads` can be influenced by `OMP_NUM_THREADS` env var or `omp_set_num_threads` at runtime.
**So both are respected** (and `omp_set_num_threads` should have priority over `OMP_NUM_THREADS` to respect openmp spec so I'm strongly against making the env var top priority).

The only thing it adds on top of openmp default behavior is an upper bound on the number of threads (`joblib.cpu_count`). It's useful in some situations like in containers because `omp_get_max_threads` might over estimate the number of available cpus. On the other hand if someone wants to use more threads than cpus, it will be capped to the number of cpus.

So the question is just do we want to bound by the number of cpus and my opinion is yes because the disadvantage seems negligible to me.

jeremiedbb committed:
Merge branch 'master' into openmp-effective-njobs

jeremiedbb committed:
make module private

jeremiedbb commented:
> "I know what I'm doing and I want that many threads no matter what anybody else says".

If we want to have that we can implement it in sklearn config as @thomasjpfan proposed. I suggest to do so in another PR

thomasjpfan commented:
> So the question is just do we want to bound by the number of cpus and my opinion is yes because the disadvantage seems negligible to me.

I am happy with bounding the number of cpus.

> If we want to have that we can implement it in sklearn config as @thomasjpfan proposed. I suggest to do so in another PR

My suggestion was for a way to "unbound it" if a user really wants to do it. I agree with discussing this in another PR to keep this PR moving along.

jeremiedbb commented:
>  I'm strongly against making the env var top priority

I can be more specific on that
```
OMP_NUM_THREADS=4 python
> run some code with 4 threads
> # locally want to use 1 thread
> with omp_set_num_threads(1):
>     run some code with 1 thread
> run some code with 4 threads
```

NicolasHug commented:
```
OMP_NUM_THREADS=4 python
> run some code with 4 threads
> # locally want to use 1 thread
> with omp_set_num_threads(1):
>     run some code with 1 thread
> run some code with 4 threads
```


We should never write code like `with omp_set_num_threads(1):` in scikit-learn

NicolasHug commented:
> during the last meeting. We went for option 5. But we allowed to keep the discussion opened on a case by case. It does not mean we won't use 5 everywhere :)

I don't remember this being agreed on (not saying it didn't happen, I might just have missed it.)

Having estimator expose n_jobs for openmp while some others won't is going to be:

- terribly hard to explain
- terribly hard for users to understand

Joblib in itself is already fairly complicated. Scikit-learn + Joblib is another beast. It took me like 5 hours to come up with some decent documentation in #15116 to describe Scikit-learn + Joblib + simple openMP.

Now we want to have Scikit-learn + Joblib + complex OPENMP + ( Joblib * complex OPENMP)??


This is going to be bad for us.

jeremiedbb commented:
Not insiste sklearn. But nothing should prevent me from doing that when using sklearn.

NicolasHug commented:
Regarding OMP_NUM_THREADS:

If the users sets OMP_NUM_THREADS manually, scikit-learn should respect it.

That's how other libraries like lightGBM and XGBoost work.

Doing anything else will only add confusion, for everyone.

jeremiedbb commented:
> If the users sets OMP_NUM_THREADS manually, scikit-learn should respect it.

I never said it won't. Please read my comments [here](https://github.com/scikit-learn/scikit-learn/pull/14196#issuecomment-539078194) and [here](https://github.com/scikit-learn/scikit-learn/pull/14196#issuecomment-541114428)` I keep repeating that the env var is respected.
What I say is that at runtime you can set `omp_set_num_threads` and it will have priority over `OMP_NUM_THREADS`. It's how openmp works and there's absolutely no reason to bypass that (why would want to remove the possibility to control the number of threads at runtime ?).

> That's how other libraries like lightGBM and XGBoost work.

Others doing bad things is not a reason to do so ourselves. Anyway it's not the case because they also respect `omp_set_num_threads`.
```
OMP_NUM_THREADS=2 python
> LGBMClassifier().fit(X, y)  # 2 threads
> with threadpool_limits(limits=1):  # calls omp_set_num_threads
>    LGBMClassifier().fit(X, y)  # 1 thread
```

The only thing this PR adds is to limit it to the number of cpus to avoid oversubscription issues in containers (which lightgbm and xgboost don't do). So yes if you have 4 cpus and set `OMP_NUM_THREADS=400` it will only use 4 threads, but that does not seem a bad thing to me.

NicolasHug commented:
> What I say is that at runtime you can setomp_set_num_threadsand it will have priority overOMP_NUM_THREADS`. It's how openmp works and there's absolutely no reason to bypass that (why would want to remove the possibility to control the number of threads at runtime ?

I am completely OK with that. 

The problem I have is with

> Current behavior is respect OMP_NUM_THREADS unless it exceeds joblib.cpu_count.

You can't say that and then claim that `OMP_NUM_THREADS` is respected.

tomMoral commented:
The point is does it makes sense to cap `OMP_NUM_THREADS` to the number of CPUs or not. To summarize, the 2 points:

- **Against** : some informed people might not understand why we do it as oversubscription might help in some cases and setting this limit will be harder to explain.

- **For** : this limits is defined a performance limit and it will avoid catastrophic behaviors with oversubscription. It is also enforced by MKL and OpenBLAS.

I am in favor of setting this limit because I have not seen a situation where oversubscription helped and it can lead to catastrophic behaviors. Also, I think the documentation issue should not influence the choice.

NicolasHug commented:
> It is also enforced by MKL and OpenBLAS.

Can you expand a bit please? That seems contradictory with what we say in https://github.com/scikit-learn/scikit-learn/pull/15116:

 - Manually setting one of the environment variables (``OMP_NUM_THREADS``,
   ``MKL_NUM_THREADS``, ``OPENBLAS_NUM_THREADS``, or ``BLIS_NUM_THREADS``)
   will take precedence over what joblib tries to do

> the documentation issue should not influence the choice

I strongly disagree here. If we can't explain it properly, users won't understand it. There's not point in providing software that users cannot understand. And I'm not even talking about maintenance cost that comes with complex designs.

tomMoral commented:
> > It is also enforced by MKL and OpenBLAS.
> 
> Can you expand a bit please? That seems contradictory with what we say in #15116:

In my understanding, for MKL and OpenBLAS, the environment variables are caped by the number of CPUs. The following script, it should not use 100 cpus but the maximal number of cpus available on your machine:
```
OMP_NUM_THREADS=100 python -c "import numpy as np; A = np.random.rand(10000, 10000); A.dot(A.T)"
```
but @jeremiedbb  knows better than me in this matter. This means that we would copy the behavior from `numpy`, which was the original goal.

> > the documentation issue should not influence the choice
> 
> I strongly disagree here. If we can't explain it properly, users won't understand it.

I agree that it is necessary to be able to explain our API choice but it seems a bit much to say that we won't be able to explain that overall, the maximal number of threads is caped by the number of CPUs.

jeremiedbb commented:
Thanks @NicolasHug for clarifying your concern. We weren't arguing on the same thing the whole time :)

> In my understanding, for MKL and OpenBLAS, the environment variables are caped by the number of CPUs.

That's right. If you set `MKL_NUM_THREADS` or `OPENBLAS_NUM_THREADS`, the max number of threads they'll use will still be capped by a maximum value (#cpus for openblas, #physical cores for mkl).

If you set `OMP_NUM_THREADS`, mkl will respect that up to it's max value. That's what is replicated in the PR.

> Now we want to have Scikit-learn + Joblib + complex OPENMP + ( Joblib * complex OPENMP)??

the fact that we use a joblib function in this helper does not mean it has an interaction with joblib parallelism tools. To bound by number of cpu we need a way to find that number. Moreover we want a method clever enough for docker-like situations. It happens that joblib already has one so we're better using it instead of re-implement it ourselves. So about the documentation we only need to say that `OMP_NUM_THREADS` is respected up to the number of cpus (no need to involve joblib here).

jeremiedbb committed:
revert ht eunintended revert  |  cln

NicolasHug commented:
Sorry for the confusion

Capping `OMP_NUM_THREADS` to the number of CPUs is one thing. It seems natural, especially since other libraries do that as well.

But capping `OMP_NUM_THREADS` to the number of available CPUs (to avoid problems in docker etc.) is another thing. This is where I'm concerned: that's a good thing in general, but my point is that if the user manually sets `OMP_NUM_THREADS`, we should not enforce this clever limit. Users expect their choice to be respected here. And it makes benchmarking against other libraries (e.g. LightGBM) impossible in some cases, since they will use a different effective number of threads.

This is also  @rth 's concern, as far as I can tell:

> My concerns is more in the case loky limits the number of cores for some reason (say you are inside a parallel process), then no mater what you set as OMP_NUM_THREAD on CLI it will have no effect, which is somewhat counter-intuitive. Also it's much easier to explain that you can always force the number of threads manually with OMP_NUM_THREADS , as opposed to some complicated logic inside loky.

----

Regarding this

> Now we want to have Scikit-learn + Joblib + complex OPENMP + ( Joblib * complex OPENMP)??

I was referring here to having special cases for handling pranged estimators, instead of using the uniform way advertised in  #15116. But that's another issue which is independent of the one above. I'm OK with discussing this one when/if the time comes.

tomMoral commented:
> Capping `OMP_NUM_THREADS` to the number of CPUs is one thing. It seems natural, especially since other libraries do that as well.
> 
> But capping `OMP_NUM_THREADS` to the number of available CPUs (to avoid problems in docker etc.) is another thing. 

Could you clarify the difference? This seems to be the same for me. If you have a virtual machine like docker, the config of the machine gives you a number of CPUs given by `cpu_count`.

rth commented:
> In my understanding, for MKL and OpenBLAS, the environment variables are caped by the number of CPUs.

They may be but I'm not sure it's part of the OpenMP spec. For instance, libgomp does not seem to limit the number of threads https://gcc.gnu.org/onlinedocs/libgomp/OMP_005fTHREAD_005fLIMIT.html#OMP_005fTHREAD_005fLIMIT 

So doing something like,
```py
if os.environ.get("OMP_NUM_THREADS"):
    # fall back to user provided number of threads
	max_n_threads = openmp.omp_get_max_threads()
else:
	max_n_threads = min(openmp.omp_get_max_threads(), cpu_count())
```
would be less disruptive I think. Whether there are cases where it would be useful is besides the point, but rather that we should avoid silent and surprising behavior. Personally I don't read the docs for BLIS, OpenBLAS and MKL in detail, but I know that setting `OMP_NUM_THREADS` will set the number of threads. If there is some logic limiting it in the case where it's manually provided, IMO we should leave it to those libraries, rather than re-implementing some custom solution ourselves.

Of course this sidesteps the fact that each of those libraries also have their own ENV variable to control this, but well..

jeremiedbb commented:
> Basically: capping to 32 makes sense. But we should not cap to 2 when the user sets OMP_NUM_THREADS=4.

I don't see the difference between
"we should cap to 32 when the user sets OMP_NUM_THREADS=34."
and
"But we should not cap to 2 when the user sets OMP_NUM_THREADS=4."

In a docker image with 2 cpus, the fact that openmp sees 32 cpus is not a feature but a defect.
In that case it makes no sense to cap to the number of cpus in the first case either.

> They may be but I'm not sure it's part of the OpenMP spec. For instance, libgomp does not seem to limit the number of threads

Indeed it does not. It's the whole purpose of this helper.

> So doing something like,
   if os.environ.get("OMP_NUM_THREADS"):

What if I did not set the env var, but I changed the number of threads with `omp_set_num_threads` ? Why the env var should be respected at all cost and not the programmatic way ? The issue is that there's no way to check if I changed the number of threads using the function.

> Personally I don't read the docs for BLIS, OpenBLAS and MKL in detail, but I know that setting OMP_NUM_THREADS will set the number of threads

Yes but it will be caped to the number of cpus (or physical cores for mkl). And it's preferred to use MKL_NUM_THREADS (or equivalent)

>  If there is some logic limiting it in the case where it's manually provided, IMO we should leave it to those libraries, rather than re-implementing some custom solution ourselves.

Now that we do code with openmp we are one of those libraries, just like mkl. So we need to implement some custom function.

If you want we can have a SKLEARN_OMP_NUM_THREADS env var which has priority over OMP_NUM_THREADS, just like mkl. If a user sets it it will be the number of OpenMP threads no matter the number of cpus (not like mkl). But please show me an example where it's really worth to use more threads than cpus with scikit-learn.

NicolasHug commented:
> But please show me an example where it's really worth to use more threads than cpus with scikit-learn

Probably none. But that's not the point. The point is to respect what users want. We don't want scikit-learn to be an outsider in the way it treats `OMP_NUM_THREADS`.

rth commented:
> But please show me an example where it's really worth to use more threads than cpus with scikit-learn

One use case could be when either 
 - CPython's `cpu_count` is wrong on some platform: it happened before e.g. https://bugs.python.org/issue30581 or https://bugs.python.org/issue33166, could happen again with new or exotic hardware
 - or loky.cpu_count is wrong in some specific conditions. I contributed the initial PR for it, and I can't guarantee that there is no platform or conditions under which it could behave differently than what we would expect.

So by clipping user provided `OMP_NUM_THREADS` we are saying we trust those two things more than we trust users their understanding of their hardware.

Another use case is simply to evaluate the behavior of the system under over-subscription (e.g. on a multi-tenant or multi-process system). It using 2 threads per CPU cores produces 10x slowdown (related to https://github.com/scikit-learn/scikit-learn/issues/15078) it's interesting to know and probably means that we don't understand something and it could be worth investigating.

A third use case, is one runs on some VM environment one don't fully understand (e.g. cloud provider) and wants to see what is the best number of threads, so they increase `OMP_NUM_THREADS` progressively. It is expected that the run time would reach an optimum at some value. With the current approach they won't see the slowdown due to oversubsription because the number of threads will be limited at some point by this mechanism. It it's probably right, but could be wrong and they will never know because they don't have access to raw data. Also it makes more complex to understand what is going on.   

> What if I did not set the env var, but I changed the number of threads with omp_set_num_threads ? Why the env var should be respected at all cost and not the programmatic way ? 

Because it takes ~5 seconds to prepend an environment variable to any script (accounting for misspellings) and look at the result. While there are less people who know about the existence of `omp_set_num_threads`, let alone who use it. Though I agree that inconsistency is not ideal. 

> If you want we can have a SKLEARN_OMP_NUM_THREADS env var which has priority over OMP_NUM_THREADS, just like mkl. 

I mostly use OMP_NUM_THREADS for everything, because I don't really want to think about the backend. What's the advantage of a separate env variable, short of if we want to specify the parallelism level on for those few particular scikit-learn routines?

As a side note this PR really received way too many comments for the few lines it adds. Maybe in such cases a quick call over a cup of coffee/tee would sort things out faster :)

jeremiedbb commented:
> The point is to respect what users want. We don't want scikit-learn to be an outsider in the way it treats OMP_NUM_THREADS.

@NicolasHug I disagree with this point.
First all libraries using openmp don't have the same behavior. MKL limits to number of cpus, lgbm doesn't.
Then we already differ from what xgboost and lightgbm does regarding openmp threads. They both expose a n_jobs parameter which we agreed not to do. Moreover both have different defaults: xgboost has n_jobs=1 and lgbm has n_jobs=-1 by default.
Finally as I said before, doing something just because others do it should not enter into consideration. Should we put a n_jobs in histgbt just because they do ?

@rth thanks for the detailed use cases. I agree that letting more flexibility to the user could be a good thing. My concern is that in what you proposed only the env var works, not the function.

> Because it takes ~5 seconds to prepend an environment variable to any script (accounting for misspellings) and look at the result.

it does not take more time to do `from threadpoolctl import threadpool_limits; threadpool_limits(1, 'openmp')` :)
And I find it even easier and faster to be able to test the number of threads in interactive mode. I would really not cut that possibility.

>  While there are less people who know about the existence of omp_set_num_threads, let alone who use it.

as of today. Tomorrow billions of people will use threadpoolctl :D 

> I mostly use OMP_NUM_THREADS for everything, because I don't really want to think about the backend. 

It's not the recommended way. do it at your own risks ! :)

> Maybe in such cases a quick call over a cup of coffee/tee would sort things out faster :)

Let's do that.

NicolasHug commented:
I'm sorry but whether XGBoost or LightGBM expose n_jobs is completely irrelevant to the discussion here.

I'm a user, I want to benchmark XGBoost, LightGBM and sklearn. I expect `OMP_NUM_THREAD=4 bench.py` to work as expected: tell all libraries to use 4 threads. You're making this impossible.

Yes, I understand that other libraries like MKL will cap to the number of CPUS: 1) whether they should is debatable, 2) it is still a much more reasonable default than what we are about to introduce here, and 3) MKL isn't comparable to sklearn in terms of abstraction level.

Again, I'm not against capping to the number of actually available CPUs: this is a good **default** to have. But users should be able to bypass that default. That's what a default is for. I don't understand why you are so keen on forbidding users to use what they want to use. On top of  the reasons @rth mentioned and mine, there probably are many others that we don't know about yet. We don't know everything.

That's all I have on this.

jeremiedbb commented:
> I don't understand why you are so keen on forbidding users to use what they want to use. 

@NicolasHug you missed one thing in my last comment. I actually agreed that we can let some flexibility. In fact I never really was against that. What I'm really against since the beginning is that it can only be controlled by the OMP_NUM_THREADS env var and not by the omp_set_num_threads function.

> I'm sorry but whether XGBoost or LightGBM expose n_jobs is completely irrelevant to the discussion here.

It's relevant in the sense that you're telling that we should do as they do. But we don't do as they do for some openmp related behavior (e.g. n_jobs) so what is special about this one (env var) ?

ogrisel commented on sklearn/utils/_openmp_helpers.pyx:
```diff
@@ -0,0 +1,38 @@
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between the number of cpus
+      and ``openmp.omp_get_max_threads()``.
```
This should be rephrased to be more precise: "the number of cpus taking ``cgroups`` quotas into account  and ``openmp.omp_get_max_threads()``. ``cgroups`` quotas can typically be set by tools such as Docker."

Note that ``openmp.omp_get_max_threads()`` already takes number of CPUs (including SMT) and CPU affinity mask constraints into account by default for all OpenMP implementations as far as I know (I have not checked though). So there is no need to mention those here.

ogrisel reviewed (commented):
Alright so the consensus is to implement something like https://github.com/scikit-learn/scikit-learn/pull/14196#issuecomment-541745512 so that we give more control to the user if the env var is explicitly set? I agree with Jeremie that the env var is not so special compared to some code that would explicitly `set_omp_num_threads` but `get_omp_num_threads` will not tell us if this was set explicitly programmatically or if the value stems from the default CPU detection which is broken with cgroups.

So +1 to not apply the `cpu_count()` cap when the env var is set.

@jeremiedbb please feel free to update this PR.

jeremiedbb committed:
Merge branch 'master' into openmp-effective-njobs

jeremiedbb committed:
allow more user control + more comments

jeremiedbb committed:
fix indent

jeremiedbb committed:
import os :)

ogrisel reviewed (approved):
Small nitpicks, but otherwise LGTM.

ogrisel commented on sklearn/utils/_openmp_helpers.pyx:
```diff
@@ -0,0 +1,48 @@
+    IF SKLEARN_OPENMP_SUPPORTED:
+        if os.getenv("OMP_NUM_THREADS"):
+            # Fall back to user provided number of threads making it possible
+            # to exceed the number of cpus.
```
Hum, ok indeed... Let's keep this comment as it is then.

jeremiedbb committed:
address comments

jeremiedbb commented on sklearn/utils/_openmp_helpers.pyx:
```diff
@@ -0,0 +1,48 @@
+            # Fall back to user provided number of threads making it possible
+            # to exceed the number of cpus.
+            # It's however inconsistent with `omp_set_num_threads` which can't
+            # be used this purpose.
```
ok, but I do it unwillingly :unamused:

jeremiedbb commented on sklearn/utils/_openmp_helpers.pyx:
```diff
@@ -0,0 +1,46 @@
+IF SKLEARN_OPENMP_SUPPORTED:
+    import os
+    cimport openmp
+    from . import cpu_count
```
yep, forgot that...

jeremiedbb committed:
better docstring

NicolasHug reviewed (approved):
LGTM, thanks @jeremiedbb
